### PR TITLE
Chapter 5

### PR DESCRIPTION
**Pull request by dibyendumajumdar**
In Chapter 5 we add If statement

### CODE REVIEW

**Comment by XmiliaH** [comment]
Currently `if` block seem to not create new scopes around its statements. The following code is accepted and results in `return 2`.
```c
if (arg == 1)
     int b = 2;
return b;
```

**Comment by dibyendumajumdar** [comment]
> Currently `if` block seem to not create new scopes around its statements. The following code is accepted and results in `return 2`.
> 
> ```c
> if (arg == 1)
>      int b = 2;
> return b;
> ```

Its a problem in grammar I think, declaration should not be allowed here. I checked that in Java you can't write this, its an error.

We have block statements which do create scope.
So

```
if (arg == 1)
{
   int b = 1;
}
```

Would work as expected I think.

**Comment by XmiliaH** [comment]
Yes that is one way to fix this issue.

As for merging. Shouldn't it be much simpler? After the `if` both scopes merged have equal `_scopes`, since no new variables can be added to outer scopes and no can be removed. It also followes that the indices will be the same. Since `$ctrl` is also in slot 0 it could be as simple as:
```java
public final void merge(RegionNode region, ScopeNode ifScope) {
	assert _scopes.equals(ifScope._scopes);
	assert nIns() == ifScope.nIns();
	for (int i = 1; i < nIns(); i++) {
		var ifTdef = ifScope.in(i);
		var ifFdef = in(i);
		if (ifTdef != ifFdef) {
			set_def(i, new PhiNode(region, ifTdef, ifFdef).peephole());
		}
	}
}
```

**Comment by dibyendumajumdar** [comment]
> Yes that is one way to fix this issue.
> 
> As for merging. Shouldn't it be much simpler? After the `if` both scopes merged have equal `_scopes`, since no new variables can be added to outer scopes and no can be removed. It also followes that the indices will be the same. Since `$ctrl` is also in slot 0 it could be as simple as:
> 
> ```java
> public final void merge(RegionNode region, ScopeNode ifScope) {
> 	assert _scopes.equals(ifScope._scopes);
> 	assert nIns() == ifScope.nIns();
> 	for (int i = 1; i < nIns(); i++) {
> 		var ifTdef = ifScope.in(i);
> 		var ifFdef = in(i);
> 		if (ifTdef != ifFdef) {
> 			set_def(i, new PhiNode(region, ifTdef, ifFdef).peephole());
> 		}
> 	}
> }
> ```

Yeah, you may be right. Cliff is refactoring this so lets wait to see the outcome

**Comment by cliffclick** [comment]
Well, the rewrite looks like this, essentially identical to @XmiliaH 
```java
    public RegionNode mergeScopes(ScopeNode that) {
        RegionNode r = (RegionNode)new RegionNode(ctrl(), that.ctrl()).peephole();
        for( int i=1; i<nIns(); i++ )
            set_def(i,new PhiNode(r, in(i), that.in(i)).peephole());
        that.kill();            // Kill merged scope
        return r;
    }

```
Still cleaning up details, will be done soon

**Comment by dibyendumajumdar** [comment]
@XmiliaH let me know if you'd like to join as a contributor or member

**Comment by XmiliaH** [comment]
For now I will stick to finding issues and providing fixes if the change is not to substential.

File: chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java [hunk 1]
```diff
@@ -86,23 +95,30 @@ public Node update(String name, Node n) {
...
-            for (Map.Entry<String, Integer> e: tab.entrySet()) {
-                String name = e.getKey();
-                Integer idx = e.getValue();
-                Node n = in(idx);
-                dupScope.define(name, n); // Ensure that dupScope is a user of n
-            }
+        ScopeNode dup = new ScopeNode();
+        for( HashMap<String, Integer> tab : _scopes ) {
+            dup.push();    // Create same level in the duplicate
+            for( String name : tab.keySet() )
```
Comment (XmiliaH) [inline chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java:109]: The order of `keySet` is not the order of insertion and so might mix up the indices for merging.

The following program returns `Phi(Region15,4,2)`, but `2` is never assigned to `c`.
```c
int c = 3;
int b = 2;
if (arg == 1) {
    b = 3;
    c = 4;
}
return c;
```

Comment (dibyendumajumdar) [reply chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java:109]: The version I wrote didn't assume indices were in same order in the dup, but with the new approach, we have to make sure I guess

File: chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java [hunk 2]
```diff
@@ -19,7 +19,7 @@ public class ScopeNode extends Node {
     public final Stack<HashMap<String, Integer>> _scopes = new Stack<>();
 
     @Override
-    public String label() { return "Scope" + _nid; }
+    public String label() { return "Scope"; }
```
Comment (dibyendumajumdar) [inline chapter05/src/main/java/com/seaofnodes/simple/node/ScopeNode.java:22]: isn't this going to break the Graph Visualizer? We needed distinct labels to be able to draw separate clusters

File: chapter05/src/main/java/com/seaofnodes/simple/Parser.java [hunk 3]
```diff
@@ -479,7 +479,7 @@ private boolean isPunctuation(char ch) {
 
         private String parsePunctuation() {
             int start = _position;
-            return new String(_input, start, start+1);
+            return new String(_input, start, 1);
```
Comment (dibyendumajumdar) [inline chapter05/src/main/java/com/seaofnodes/simple/Parser.java:482]: punctuations can be more than 1 char long - e.g. ==

Comment (XmiliaH) [reply chapter05/src/main/java/com/seaofnodes/simple/Parser.java:509]: This is used for error messages where it is unclear how long the punctuation shoould be. I think using only on character here is fine.

Comment (cliffclick) [reply chapter05/src/main/java/com/seaofnodes/simple/Parser.java:509]: well, the original code mis-used the new String API, and had the position being used as the length, so I think this is a clear winner.

**Review by XmiliaH — APPROVED** [review]
Would like to see the fixed order in the graphs, but that seems to be currently an issue of graphviz.

**Review by cliffclick — APPROVED** [review]
Long and winding road, but finally done.

**Pull request merged**